### PR TITLE
Countdown component for submissions

### PR DESCRIPTION
**Pull request by meems1996**
## Description
Built a countdown component placed above the description of the Nerd Days 1.0 for Issue #674 
When the time expired the countdown stops and displays a paragraph element with the text Submissions are closed. 

## Screenshot(s)
Dark mode version of example with counter active
Light mode version of example with counter active
Dark mode version of example with counter not active, shows submissions are closed text
Light mode version of example with counter not active, shows submissions are closed text

### CODE REVIEW

File: src/components/Countdown/CountdownContainer.module.scss [hunk 1]
```diff
@@ -0,0 +1,33 @@
+.container {
+    display: flex;
+    justify-content: center;
+    padding-right:20px;
+}
+
+.countdownContainer {
+    border: 1px solid rgb(213, 215, 215);
```
[the remark this answers is not in the document]
Comment (caylahamann) [reply src/components/Countdown/CountdownContainer.module.scss:8]: @jerelmiller created an issue for this! [#90](https://github.com/newrelic/gatsby-theme-newrelic/issues/90) 😄

[the remark this answers is not in the document]
Comment (jerelmiller) [reply src/components/Countdown/CountdownContainer.module.scss:8]: Thanks @caylahamann for opening that!

File: src/components/Countdown/CountdownContainer.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  }
+
+  return (
+    <div className={styles.container}>
+      {active ? (
+        <div className={styles.countdownContainer}>
+          <Countdown countdown={countdown} />
+        </div>
+      ) : (
+        <p>Submissions are closed</p>
```
Comment (djsauble) [inline src/components/Countdown/CountdownContainer.js:63]: For reusability, can this string be defined where the component is used? I imagine this component would be handy for a number of things, not just submission deadlines.

File: src/pages/index.js [hunk 3]
```diff
@@ -65,6 +66,7 @@ const IndexPage = ({ data, pageContext }) => {
             className={cx(styles.intro, 'intro-text')}
           >
             <div className={styles.introText}>
+              <CountdownContainer />
```
Comment (djsauble) [inline src/pages/index.js:69]: The way the countdown timer is positioned on the page, I expect it to be a countdown to the event itself, not the submission deadline. I suggest moving this under the paragraph which talks about submitting a proposal.

File: src/components/Countdown/CountdownContainer.module.scss [hunk 4]
```diff
@@ -0,0 +1,33 @@
+.container {
+    display: flex;
+    justify-content: center;
+    padding-right:20px;
+}
+
+.countdownContainer {
```
Comment (djsauble) [inline src/components/Countdown/CountdownContainer.module.scss:7]: I love the border and spacing of the countdown timer. When the timer expires, it should have the same border and spacing (bonus: if you're watching it expire, it won't jump around when it changes state).

File: src/components/Countdown/CountdownContainer.js [hunk 5]
```diff
@@ -0,0 +1,69 @@
+import React, { useEffect } from 'react';
+import Countdown from './Countdown';
+import styles from './CountdownContainer.module.scss';
+
+const CountdownContainer = () => {
+  const [countdown, setCountdown] = React.useState(() => getRemainingTime());
+  const [active, setActive] = React.useState(() => {
```
[the remark this answers is not in the document]
Comment (meems1996) [reply src/components/Countdown/CountdownContainer.js:7]: Hi @jerelmiller! Yes. I am taking a look now. Thank you for your feedback! It's been super fun to work on over the weekend.

File: src/components/Countdown/CountdownContainer.js [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        setActive(false);
+        clearInterval(interval);
+      } else {
+        setCountdown({ days, hours, minutes, seconds });
+      }
+    }, 1000);
+  }, []);
+
+  function getRemainingTime() {
+    const countDownDate = new Date('September 1, 2020 23:59:59').getTime();
```
[the remark this answers is not in the document]
Comment (meems1996) [reply src/components/Countdown/CountdownContainer.js:42]: You're right! It does make a lot of sense having it be a prop!

File: src/components/Countdown/CountdownContainer.js (hunk 2 above)
Comment (meems1996) [reply src/components/Countdown/CountdownContainer.js:63]: Great idea! Thank you for your feedback @djsauble  :-)

File: src/components/Countdown/Countdown.js [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+      <h2>:</h2>
+      <CountdownUnit value={countdown.minutes} label="Minutes" />
+      <h2>:</h2>
+      <CountdownUnit value={countdown.seconds} label="Seconds" />
+    </div>
+  );
+};
+
+Countdown.propTypes = {
+  countdown: PropTypes.number,
```
[the remark this answers is not in the document]
Comment (meems1996) [reply src/components/Countdown/Countdown.js:21]: Do you mean something like this? 

```js
Countdown.propTypes = {
  countdown: PropTypes.shape({
    days: PropTypes.number,
    hours: PropTypes.number, 
    minutes: PropTypes.number,
    seconds: PropTypes.number,
  })
};
```

File: src/components/Countdown/Countdown.js [hunk 8]
```diff
@@ -0,0 +1,24 @@
+import React from 'react';
+import CountdownUnit from './CountdownUnit';
+import PropTypes from 'prop-types';
+import styles from './CountdownContainer.module.scss';
+
+const Countdown = ({ countdown }) => {
+  return (
+    <div className={styles.countdown}>
+      <CountdownUnit value={countdown.days} label="Days" />
+      <h2>:</h2>
```
[the remark this answers is not in the document]
Comment (meems1996) [reply src/components/Countdown/Countdown.js:10]: Thanks for making me aware of the use of `var(--border-color)`. What do you use for the headings and text that switches from dark to light mode? I have been trying to find it but I am new to both the codebase and scss haha

File: src/components/Countdown/Countdown.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (jerelmiller) [reply src/components/Countdown/Countdown.js:21]: Yep! That looks perfect!

File: src/components/Countdown/Countdown.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (jerelmiller) [reply src/components/Countdown/Countdown.js:10]: That's a really great question! There are a few moving pieces to this equation.

tl;dr; 

We have classes that get added to the `body` tag, and we use [CSS variables](https://developer.mozilla.org/en-US/docs/Web/CSS/Using_CSS_custom_properties) to switch the primary text color depending on which "mode" the site is in (dark or light). You can see the common global variables here: https://github.com/newrelic/gatsby-theme-newrelic/blob/develop/packages/gatsby-theme-newrelic/src/components/GlobalStyles.js#L104-L146

### Deep dive

We are using a Gatsby plugin called [`gatsby-plugin-use-dark-mode`](https://www.gatsbyjs.com/plugins/gatsby-plugin-use-dark-mode/) which, in turn uses a library called [`use-dark-mode`](https://github.com/donavon/use-dark-mode#readme). Essentially these libraries are responsible for toggling the class on the `body` tag between `light-mode` and `dark-mode` (Try it! If you inspect the body tag and click the dark mode toggle, you can see this in action 🙂 ). 

We use the `gatsby-plugin-use-dark-mode` plugin in our [`gatsby-config.js`](https://github.com/newrelic/gatsby-theme-newrelic/blob/develop/packages/gatsby-theme-newrelic/gatsby-config.js#L7) for the theme and use the `use-dark-mode` hook in the [`DarkModeToggle`](https://github.com/newrelic/gatsby-theme-newrelic/blob/develop/packages/gatsby-theme-newrelic/src/components/DarkModeToggle.js) component (which is the element you click on in the header to toggle the mode!). 

We take advantage of the `light-mode` or `dark-mode` class that gets toggled on the `body` to swap out the values of the CSS variables. If you look at our [global styles](https://github.com/newrelic/gatsby-theme-newrelic/blob/develop/packages/gatsby-theme-newrelic/src/components/GlobalStyles.js#L104-L146), you'll notice that we can use a CSS variable as the value of another CSS variable:

```css
--primary-text-color: var(--color-neutrals-700);
```

This is really powerful, because it allows us to create abstractions in our CSS. Combined with our [color](https://github.com/newrelic/gatsby-theme-newrelic/blob/develop/packages/gatsby-theme-newrelic/src/components/GlobalStyles.js#L5-L102) definitions, we can then target the mode and set the appropriate text color accordingly:

```css
.light-mode {
  --primary-text-color: var(--color-neutrals-700);
}

.dark-mode {
  --primary-text-color: var(--color-dark-700);
}
```
(source: [light mode](https://github.com/newrelic/gatsby-theme-newrelic/blob/develop/packages/gatsby-theme-newrelic/src/components/GlobalStyles.js#L107), [dark mode](https://github.com/newrelic/gatsby-theme-newrelic/blob/develop/packages/gatsby-theme-newrelic/src/components/GlobalStyles.js#L129))

Whats super cool about that is that now I can use the `--primary-text-color` variable anywhere I want "dark mode aware" text that matches the prominent text color used throughout the site. In fact, that's exactly what we do for the `body` selector by setting the [`color` property](https://github.com/newrelic/gatsby-theme-newrelic/blob/develop/packages/gatsby-theme-newrelic/src/components/GlobalStyles.js#L173) to `--primary-text-color`. I find that to be pretty cool 🎉 .

For headings, we do something very similar. As you've noticed on the site, our headings are much brighter than the primary text color. Here we've defined a `--heading-text-color` for [light mode](https://github.com/newrelic/gatsby-theme-newrelic/blob/develop/packages/gatsby-theme-newrelic/src/components/GlobalStyles.js#L122) and [dark mode](https://github.com/newrelic/gatsby-theme-newrelic/blob/develop/packages/gatsby-theme-newrelic/src/components/GlobalStyles.js#L144), which we then use to set the [heading colors](https://github.com/newrelic/gatsby-theme-newrelic/blob/develop/packages/gatsby-theme-newrelic/src/components/GlobalStyles.js#L203).

Whew, that was a lot 😄 . Hopefully that helps you get a little better understanding of some of the mechanics we've employed to pull off our dark mode theme and where to find them in the code base!

[the remark this answers is not in the document]
Comment (jerelmiller) [reply src/components/Countdown/Countdown.js:10]: > What do you use for the headings and text that switches from dark to light mode? 

Because it probably wasn't super obvious in that last comment you can use the `var(--heading-text-color)` variable to match the heading text color styles, which is dark mode aware.

File: src/pages/index.js [hunk 9]
```diff
@@ -17,6 +17,7 @@ import styles from './index.module.scss';
 import devChampionBadge from '../images/developer-champion/dev-champion-badge.png';
 import podcastBadge from '../images/podcasts/podcasts-badge.png';
 import nerdDays from '../images/nerd-days/nerd-days.png';
+import CountdownContainer from '../components/Countdown/CountdownContainer';
```
[the remark this answers is not in the document]
Comment (meems1996) [reply src/pages/index.js:20]: Thanks! This is very helpful to know!

File: src/components/Countdown/CountdownContainer.module.scss (hunk 4 above)
Comment (meems1996) [reply src/components/Countdown/CountdownContainer.module.scss:7]: Thank you! Is this better? 

Dark mode version of example with counter active
Dark mode version of example with counter active
Light mode version of example with counter active
Light mode version of example with counter active

Comment (djsauble) [reply src/components/Countdown/CountdownContainer.module.scss:7]: Looks great! 🎉

Comment (meems1996) [reply src/components/Countdown/CountdownContainer.module.scss:7]: @djsauble Thank you 😃 I just noticed you approved the PR and I just submitted some last changes. I noticed that I forgot to specify the time zone. Thank you so much for your feedback! This is my first time attempting to contribute to an open source project and it has been a wonderful experience! 🎉

**Comment by jerelmiller** [comment]
Thanks again @meems1996 for contributing to this project! That is so exciting this was your first contribution 🎉 !! We look forward to more from you!

**Comment by meems1996** [comment]
@jerelmiller Thank you as well for your feedback. I have not used PropTypes before and it's been a great learning experience contributing. I am very happy New Relic provides these opportunities. 💯
